Dynamic config for pager

jenswikl opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswikl committed:
core: fix boot_mem_release_tmp_alloc()

boot_mem_release_tmp_alloc() releases memory between the recorded
mem_end and orig_mem_end. However, if mem_start has reached the same
page as mem_end before boot_mem_release_unused() is called we're using a
part the first page in the temporary area boot_mem_release_tmp_alloc()
tries to release. So fix this by keeping track of the last used page at
the end of boot_mem_release_unused() and make sure not to free that page
in boot_mem_release_tmp_alloc().

Fixes: fe85eae5b0b3 ("core: add CFG_BOOT_MEM and  boot_mem_*() functions")
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: mm: update is_nexus()

Update the function is_nexus() to take additional NEX_RAM_RW and
NEX_DYN_VASPACE sections into account if the virtual address isn't in
the first nexus range.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>

jenswikl committed:
core: mm: add mobj_phys_alloc_flags()

Add mobj_phys_alloc_flags() to cover the use case where physical memory
is mapped at a dynamically allocated virtual address in for instance the
memory areas MEM_AREA_NEX_DYN_VASPACE and MEM_AREA_TEE_DYN_VASPACE.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: mm: add mobj_page_alloc()

Add mobj_page_alloc() as an alternative to virt_page_alloc() when a
struct  mobj is needed to represent that allocated pages.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>

jenswikl committed:
core: simplify core_init_mmu_map()

Simplify start address calculation for the dummy static_memory_map used
while initializing the translation tables.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>

jenswikl committed:
core: mm: pgt_init(): use virt_page_alloc() if available

If configured with CFG_DYN_CONFIG=y and CFG_CORE_PREALLOC_EL0_TBLS=n and
CFG_WITH_PAGER=n, allocate the translation tables using
virt_page_alloc() instead of static allocation in the .nozi.pgt_cache
section.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>

jenswikl committed:
core: arm: init_user_kcode(): use mobj_page_alloc() if available

If configured with CFG_DYN_CONFIG=y, and init_user_kcode() needs
thread_user_kdata_page, allocate it using mobj_page_alloc() instead of
static a allocation in the .nozi.kdata_page or .nex_nozi.kdata_page
section.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: arm: allocate heap from physical memory pool

For configurations without paging enabled, allocate heaps from the
physical memory pool if CFG_DYN_CONFIG=y. This avoids the static
allocations in the linker script and enables more flexible memory
management.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: tee_pager_phys_to_virt(): support ASLR and sub-page lookups

Add support for ASLR by compensating for the map-offset first in the
linear lookup and then when comparing with the non-relocated constant
TEE_RAM_START.

Add support to lookup addresses that aren't page aligned, but with the
memory range within the page.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: arm: support allocating xlat table from pager

Add support for allocating translation tables from pager once the MMU
has been enabled. This is needed in later patches CFG_DYN_CONFIG=y when
the pager is enabled.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: arm: allocate kdata page from pager

Add support for allocating the kdata page(s) from pager needed by
CFG_CORE_UNMAP_CORE_AT_EL0=y. This is needed in later patches
CFG_DYN_CONFIG=y when the pager is enabled.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: mm: core_mmu.c: add core_mmu_linear_map_end

Add core_mmu_linear_map_end, used during virt_to_phys() when
CFG_WITH_PAGER=y to keep track of the limit of linear mapped memory.
This is needed in later patches with CFG_WITH_PAGER=y and
CFG_DYN_CONFIG=y after boot memory has been released.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: thread_init_threads(): replace dummy thread ctx

When configuring paging, a thread context is needed to keep the VFP
state, but thread context might not have been allocated yet with
CFG_DYN_CONFIG=y so a dummy thread context might be needed until then.
When thread_init_threads() assigns and initializes the real thread
context, update the pointer and count variables using an unpaged
function to make sure that pager code doesn't access the thread context
while it's being replaced. This is needed in later patches with
CFG_DYN_CONFIG=y and CFG_WITH_PAGER=y.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: add boot_mem section for pager

With CFG_WITH_PAGER=y, a special section of memory is carved out by the
linker script to be used by the boot_mem*() functions during boot. The
amount of memory must cover the worst case, but the remaining unused
memory will be returned to be managed by the pager at end of boot in a
later patch. This enables boot configuration in the same way regardless
of CFG_DYN_CONFIG.

Add support in boot_mem_release_unused() and
boot_mem_release_tmp_alloc() to pass the unused memory to the pager.
This is needed in later patches with CFG_WITH_PAGER=y and
CFG_DYN_CONFIG=y.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: arm: enable dynamic config with pager

Add support for CFG_DYN_CONFIG=y and CFG_WITH_PAGER=y.  The special
.pager_boot_mem section is extended to cover the memory needed by
CFG_DYN_CONFIG=y, this includes translation tables, stacks, and heap.

Later patches can make CFG_DYN_CONFIG=y mandatory for ARCH=arm and
remove redundant code.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
core: boot_mem: remove multiple of align restriction

The two internal boot mem functions mem_alloc_tmp() and mem_alloc() has
until now required the length of the request to be a multiple of align.
While it may minimize amount of wasted padding it serves no real purpose
and with the commit bd8bea6f45b0 ("core: boot_mem: enable asan support")
the alignment is increased internally with CFG_CORE_SANITIZE_KADDRESS=y.
This makes it virtually impossible for the caller always supply correct
lengths. So fix this by removing that restriction.

Fixes: bd8bea6f45b0 ("core: boot_mem: enable asan support")
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>

jenswikl committed:
core: set uref_base for uref calculations

kaddr_to_uref() and uref_to_vaddr() was until this patch using
VCORE_START_VA as the lowest possible core data address. However, with
dynamic mapping that address might become even lower depending on the
virtual address assignment. So find out the lowest address that can be
used and use that instead of VCORE_START_VA to avoid possible integer
overflow or underflow in uref_to_vaddr() and kaddr_to_uref().

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>

etienne-lms commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1056,10 +1082,15 @@ void thread_get_user_kdata(struct mobj **mobj, size_t *offset,
 	*offset = (vaddr_t)thread_user_kdata_page -
 		  (vaddr_t)mobj_get_va(*mobj, 0, *sz);
+#endif
+	*sz = thread_user_kdata_page_size;
```
size if always `ROUNDUP(sizeof(struct thread_core_local) * CFG_TEE_CORE_NB_CORE, SMALL_PAGE_SIZE)`. Is it worth using a memory cell to store it?

jenswi-linaro replied:
In a not too distant future it will become `ROUNDUP(sizeof(struct thread_core_local) * thread_core_count, SMALL_PAGE_SIZE)`. I can wrap it in a helper function instead if you prefer that.

etienne-lms commented on core/arch/arm/kernel/boot.c:
```diff
@@ -332,7 +333,9 @@ static void init_asan(void)
 	asan_tag_access(__pageable_start, __pageable_end);
 #endif /*CFG_WITH_PAGER*/
+#ifndef CFG_DYN_CONFIG
 	asan_tag_access(__nozi_start, __nozi_end);
```
MMU tables may still be allocated from the nozi section. Shouldn't it still be tagged for ASAN access in these cases?

jenswi-linaro replied:
`__nozi_start` and `__nozi_end` doesn't exist with `CFG_DYN_CONFIG`. This is instead covered by the `boot_mem_init_asan()` call above.

etienne-lms reviewed (commented):
Some feedback. I've not cover the whole series yet and I'll need to re-read some commits. In the mean time:

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits:
"core: boot_mem: remove multiple of align restriction" and
"core: simplify core_init_mmu_map()".

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: mm: update is_nexus()",
"core: mm: add mobj_page_alloc()",
"core: mm: pgt_init(): use virt_page_alloc() if available" and
"core: set uref_base for uref calculations".

Commits that LGTM with questions or minor comments:
"core: mm: add mobj_phys_alloc_flags()", and
"core: arm: init_user_kcode(): use mobj_page_alloc() if available".

jenswi-linaro commented on core/mm/mobj.c:
```diff
@@ -196,6 +197,36 @@ struct mobj *mobj_phys_alloc(paddr_t pa, size_t size, uint32_t mem_type,
+	m->mobj.ops = &mobj_phys_ops;
+	refcount_set(&m->mobj.refc, 1);
+	m->pa = pa;
+	m->va = va;
```
Good point, I'll add a check.

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -561,6 +570,9 @@ set_core_local_kcode_offset(struct thread_core_local *cls, long offset)
 
 static void init_user_kcode(void)
 {
+	__maybe_unused struct mobj *m;
```
Yes, I'll fix.

jenswi-linaro commented:
Fixed up "core: mm: add mobj_phys_alloc_flags()" and "core: arm: init_user_kcode(): use mobj_page_alloc() if available" as
requested.

Tags applied.

jenswi-linaro commented:
Rebased on master

jenswi-linaro commented:
Rebased with fixes commit for the problems I found when testing on Hikey with pager enabled.

jenswi-linaro commented:
Rebased on master to get the CI updates.

etienne-lms commented:
Sorry for this very late feedback.
I've tested over stm32mp1 (32bit) with pager. I face issues when enabling `CFG_TEE_CORE_DEBUG`: some console trace messages are corrupted and boot hangs after pager initialization. Strangely, no issue when ``CFG_TEE_CORE_DEBUG=n` and the TZSRAM is drastically shortened. I could not find find why. I need to investigate further.

jenswi-linaro commented:
Thanks for testing.

In generic code, CFG_TEE_CORE_DEBUG enabled asserts, detailed panics, double-check virt_to_phys() and vice versa, and deadlock detection for spinlocks. The latter two seem more suspect. Perhaps you can try modifying the code to disable one of them independently?

etienne-lms commented:
Hi. I've faced issues with these changes on stm32mp15 paged target. It consumes to many internal RAM leaving few pages to the pager pool. Resulting core is very very slow. I tried to find a way to reused wasted pages but failed... and missed to notify you at the time.
Sorry for this very very late feedback.

jenswikl commented:
So it works then? Even if it's not usable on your platform. How about trying to optimize memory usage in a later PR? Do you know how many physical pages became unavailable for paging?

Another option is to forget about dynamic configuration for a pager setup and instead focus on isolating the pager setup from the dynamic configuration.

jenswikl commented:
@etienne-lms, as agreed during the OP-TEE Contributions Forum.
Rebased and squashed in the updates. There are a few commits that are missing review tags.

With this, `CFG_DYN_CONFIG` will be enabled by default for all arm configurations. However, it can still be configured with `CFG_DYN_CONFIG=n` if needed. We will not make `CFG_DYN_CONFIG=y` mandatory until all relevant platforms work with it.

etienne-lms commented:
Keep alive.
